pkgs/development/compilers/cudatoolkit/auto-add-opengl-runpath-hook.sh

```diff
@@ -1,14 +1,33 @@
 # Run autoOpenGLRunpath on all files
 echo "Sourcing auto-add-opengl-runpath-hook"
 
-autoAddOpenGLRunpathPhase  () {
-    # TODO: support multiple outputs
-    for file in $(find ${out,lib,bin} -type f); do
-        addOpenGLRunpath $file
+autoAddOpenGLRunpathPhase() (
+  containsAny() {
+    local haystack="$1"; shift;
+    for needle in "$@"; do
+      if [[ $haystack == *"$needle"* ]]; then
+        return 0
+      fi
     done
-}
+    return 1
+  }
+
+  local openGLRunpathLibs=(${openGLRunpathLibs[*]-libcuda.so libnvidia-ml.so})
+  local outputPaths=($(for o in $(getAllOutputNames); do echo "${!o}"; done))
+  find "${outputPaths[@]}" -type f -executable -print0 | while IFS= read -rd "" f; do
+    if isELF "$f"; then
+      if containsAny "$(patchelf --print-needed "$f" 2>/dev/null || :)" "${openGLRunpathLibs[@]}"; then
```

Comment:
The concern here is that -- IIRC -- we have previously encountered situations in which a library attempts to load `libcuda.so` (or others) but does not include it in DT_NEEDED. Not sure if there's a convenient resolution for that. 

I suppose it could be argued that manual intervention _should_ be required in those cases. But for whatever reason I don't think that's the approach that we've taken so far.

Perhaps @SomeoneSerge @ConnorBaker might have other thoughts?

Reply:
Yeah I suspected that would be the case. `autoPatchelf` is much stricter about this and requires that NEEDEDs be configured correctly for the auto-patching behavior. I think that's certainly the *ideal* place we should be, but it might be too painful to go there *right now*. I figured I put the ideal solution up first and and we could walk our way back from there as needed.

Reply:
The one "saving grace" we may be able to leverage is the fact that `autoAddOpenGLRunpathHook` is actually only used in about a dozen places, and most of them are in cudaPackages itself.

Reply:
> The one "saving grace" we may be able to leverage is the fact that autoAddOpenGLRunpathHook is actually only used in about a dozen places, and most of them are in cudaPackages itself.

Indeed this is convenient. However the main challenge is that we don't (yet) have a great way to test for failures... the sneaky ones all occur at runtime and require a GPU, something that nixpkgs checkPhase isn't currently set up for. Solutions to this are being discussed in https://github.com/NixOS/nixpkgs/issues/225912.

For the scope of this PR, I propose we as @NixOS/cuda-maintainers establish a policy as to whether autoAddOpenGLRunpathHook should respect DT_NEEDED or not, and then we can proceed from there. Waiting on https://github.com/NixOS/nixpkgs/issues/225912 could take a while.

Reply:
Excellent! Thank you. I'm happy to adjust this PR for expediency (likely by only filtering on a successful call to `patchelf --print-needed` so as to filter out statically linked exes).

Reply:
FWIW, *some* tests might be doable by adding `linuxPackages.x11-nvidia` to `nativeCheckInputs`. I've had some success with that, but it's likely not sure-fire.

Reply:
@samuela I agree that manual intervention should be required in those cases -- is it fair to call it an upstream bug in the way these binaries are produced? Ideally if we enumerate the places where a library is trying to load a binary not included in `DT_NEEDED`, we could report it upstream. @SomeoneSerge what are your thoughts?

@de11n thank you for your work on this :)

Reply:
I agree with @samuela that we should expect cuda libraries to dlopen() libcuda even without it being listed in the header. Minimizing the number of patching we do sounds nice, but I also agree we need to introduce automated gpu tests before we start with this kind of polishing.

Is my understanding correct that the primary motivation in this PR is to address build failures, and that these failures are caused by the hook attempting to patch executables with empty .interp field=statically linked executables?

Reply:
RE: libcuda not in DT_NEEDED being a bug

Dt_needed has an extra effect that a needed soname is loaded at the process startup (or at dlopen of the depending library). E.g. if nvidia-smi had libcuda.so as needed you couldn't run `nvidia-smi -h` on a system without the driver.

It'd be nice though to know in advance what an executable might try to load. E.g. if this info were included in the nvidia.s manifests...

Reply:
> Dt_needed has an extra effect that a needed soname is loaded at the process startup (or at dlopen of the depending library). E.g. if nvidia-smi had libcuda.so as needed you couldn't run nvidia-smi -h on a system without the driver.

Ooh that's a good point I had not previously considered. I imagine that folks might eg maintain a shared cache with `cudaSupport = true` but then run the same software on non-GPU machines. 

> It'd be nice though to know in advance what an executable might try to load. E.g. if this info were included in the nvidia.s manifests...

I suppose we could modify the hook to patch a more specific set of libraries, something like `openGLHookPatchThese = [ "foo.so", "bar.so.*" ];`. That would be akin to the manual `addOpenGLRunpath` function, but could be made more ergonomic with globbing, DT_NEEDED detection, and so forth.

Reply:
I was just about to propose the exact same idea. It would be nice to have *some* control over what this hook patches. A a regex/glob that defaults to `*` would at least let me limit it without having to write my own loops with `addOpenGLRunpath`.

Reply:
I think we can also start by testing binaries whether they're dynamically linked instead. From what I gather that should address the failures already

Reply:
Ok. I'll change it to only do that.

Reply:
I've made that change.